api/v1alpha4/cluster_webhook_test.go

```diff
@@ -46,58 +48,282 @@ func TestClusterDefault(t *testing.T) {
+	// Enabling the feature flag temporarily for this test.
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterManagedTopologies, true)()
+
+	tests := []struct {
+		name      string
+		in        *Cluster
+		old       *Cluster
+		expectErr bool
+	}{
+		{
+			name:      "should return error when topology does not have class",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{},
+				},
+			},
+		},
+		{
+			name:      "should return error when topology does not have valid version",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "invalid",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error when duplicated MachineDeployments names exists in a Topology",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "v1.19.1",
+						Workers: &WorkersTopology{
+							MachineDeployments: []MachineDeploymentTopology{
+								{
+									Name: "aa",
+								},
+								{
+									Name: "aa",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:      "should pass when MachineDeployments names in a Topology are unique",
+			expectErr: false,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "v1.19.1",
+						Workers: &WorkersTopology{
+							MachineDeployments: []MachineDeploymentTopology{
+								{
+									Name: "aa",
+								},
+								{
+									Name: "bb",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error on create when both Topology and control plane ref are defined",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					ControlPlaneRef: &corev1.ObjectReference{},
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "v1.19.1",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error on create when both Topology and infrastructure ref are defined",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{},
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "v1.19.1",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error on upgrade when Topology class is changed",
+			expectErr: true,
+			old: &Cluster{
+				Spec: ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{},
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "v1.19.1",
+					},
+				},
+			},
+			in: &Cluster{
+				Spec: ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{},
+					Topology: &Topology{
+						Class:   "bar",
+						Version: "v1.19.1",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error on upgrade when Topology version is downgraded",
```

Comment:
Maybe also add a test that upgrade works.

Reply:
Done